pkg/auth/requests/impersonate.go

```diff
@@ -78,23 +101,42 @@ func (h *impersonatingAuth) Authenticate(req *http.Request) (k8sUser.Info, bool,
 		} else {
 			groups = nil
 		}
+		if impersonateExtras {
+			extras = reqExtras
+		}
 		groups = append(groups, k8sUser.AllAuthenticated)
+	} else {
+		extras = userInfo.GetExtra()
 	}
 
-	extra := userInfo.GetExtra()
-
 	return &k8sUser.DefaultInfo{
 		Name:   user,
 		UID:    user,
 		Groups: groups,
-		Extra:  extra,
+		Extra:  extras,
 	}, true, nil
 }
 
-func groupsEqual(group1, group2 []string) bool {
-	if len(group1) != len(group2) {
-		return false
+func impersonateExtrasFromHeaders(headers http.Header) map[string][]string {
+	extras := make(map[string][]string)
+	for headerName, values := range headers {
+		if !strings.HasPrefix(headerName, authenticationv1.ImpersonateUserExtraHeaderPrefix) {
+			continue
+		}
+		extraKey := unescapeExtraKey(strings.ToLower(headerName[len(authenticationv1.ImpersonateUserExtraHeaderPrefix):]))
+		if extras == nil {
+			extras = make(map[string][]string)
+		}
+		extras[extraKey] = values
 	}
 
-	return sets.NewString(group1...).Equal(sets.NewString(group2...))
+	return extras
+}
+
+func unescapeExtraKey(encodedKey string) string {
+	key, err := url.PathUnescape(encodedKey) // Decode %-encoded bytes.
+	if err != nil {
+		return encodedKey // Always record extra strings, even if malformed/unencoded.
```

Comment:
It might be valid if it was not enconded. I changed the comment. To be honest I took the code from https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/filters/impersonation.go#L204-L210 and forgot to change the comment :sweat: